From 83746fe5d1741f8d583267a093f826406bd688d3 Mon Sep 17 00:00:00 2001
From: Banka Sai Krishna <quic_banksaik@quicinc.com>
Date: Fri, 30 Aug 2024 11:41:57 +0530
Subject: [PATCH] hostapd: Fix TPE ext duplicate IEs

When all subbands have the same PSD values, TPE EXTs are being filled
for all subbands. Address this issue by filling TPE EXTs for all
subbands only if the PSD values differ. If not, fill with a single
PSD value.

Change-Id: Ie54f6b856367f03515978f30cd1b4c8f849ab8a3

Signed-off-by: Banka Sai Krishna <quic_banksaik@quicinc.com>
---
 src/ap/ieee802_11.c | 22 ++++++++++++----------
 1 file changed, 12 insertions(+), 10 deletions(-)

--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7853,16 +7853,6 @@ static int get_psd_values(struct hostapd
 		if (!is_different_psd && (start_chan_psd != chan_psd))
 			is_different_psd = 1;
 	}
-
-	if (!is_different_psd && !punct_bitmap) {
-		*tx_pwr_count = 0;
-	} else {
-		*tx_pwr_count = num_psd_values_to_psd_count(non_11be_chan_count);
-		if (*tx_pwr_count == 0) {
-			wpa_printf(MSG_ERROR, "Invalid channel count:%d", non_11be_chan_count);
-			return -1;
-		}
-	}
 #ifdef CONFIG_IEEE80211BE
 	/* For 11be the TPE extension parameter added if the bw is 320MHZ or if
 	 * any channel is punctured in 320MHZ/160MHZ/80MHZ
@@ -7887,8 +7877,21 @@ static int get_psd_values(struct hostapd
 		}
 		tx_pwr_ext_array++;
 		*tx_pwr_ext_count += 1;
+		if (!is_different_psd && (start_chan_psd != chan_psd))
+			is_different_psd = 1;
 	}
 #endif
+	if (!is_different_psd && !punct_bitmap) {
+		*tx_pwr_count = 0;
+		*tx_pwr_ext_count = 0;
+	} else {
+		*tx_pwr_count = num_psd_values_to_psd_count(non_11be_chan_count);
+		if (*tx_pwr_count == 0) {
+			wpa_printf(MSG_ERROR, "Invalid channel count:%d", non_11be_chan_count);
+			return -1;
+		}
+	}
+
 	return 0;
 }
 
